008-dplyr-retain-and-exclude.Rmd

```diff
@@ -0,0 +1,721 @@
+
+> Drop rows with a `deceased` patient.
+
+```{r}
+# Wrong, want to keep `NA`s where `deceased` status is unknown
+patients |>
+  filter(!deceased)
+```
+
+These all work, with `exclude()` being a clear winner:
+
+```{r}
+patients |>
+  filter(!deceased | is.na(deceased))
+
+patients |>
+  filter(!(deceased %in% TRUE))
+
+# Or
+patients |>
+  filter(deceased %in% c(FALSE, NA))
+
+patients |>
+  exclude(deceased)
+```
+
+### Excluding rows using `this | that`
+
+Let's look at another example.
+Our goal is:
+
+> *Exclude* rows where class is "suv" *or* mpg is less than 15.
+
+```{r}
+cars <- tibble(
+  class = c("suv", "suv", "suv", "coupe", "coupe", "coupe", NA, NA, NA),
+  mpg = c(10, 20, NA, 10, 20, NA, 10, 20, NA)
+)
+
+cars
+```
+
+Because dplyr doesn't have a way to exclude rows, you'd reach for `filter()`.
+You'd probably first try to translate the problem statement directly to code and then invert it with a `!` in the front:
+
+```{r}
+cars |>
+  filter(!(class == "suv" | mpg < 15))
+```
+
+Hmm, we've got two issues here:
+
+-   We're again facing a scenario where we need to add boolean operators into the mix.
+
+-   We're again mistakenly dropping rows with `NA`s where we don't *know* the conditions were true.
+
+A correct `filter()` solution would look like:
+
+```{r}
+cars |>
+  filter(
+    !((class == "suv" & !is.na(class)) |
+      (mpg < 15 & !is.na(mpg)))
+  )
+```
+
+An equally correct solution would be to recognize that you can split this into two filter statements using [De Morgan's Laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws).
+
+```{r}
+cars |>
+  filter(class != "suv" | is.na(class)) |>
+  filter(mpg >= 15 | is.na(mpg))
+```
+
+This is simpler, but has *a lot* of mental overhead.
+Remember that the original problem statement was:
+
+> *Exclude* rows where class is "suv" *or* mpg is less than 15.
+
+To achieve this we've had to:
+
+-   Flip the conditions (taking special care that `< 15` is now `>= 15`)
+-   Specially handle missing values
+-   Introduce `|`, at a minimum, or `!`, `()`, and `&` if not flipping the conditions
+
+Here's the `exclude()` solution:
+
+```{r}
+cars |>
+  exclude(class == "suv" | mpg < 15)
+```
+
+Note how this *precisely* translates the problem statement into code.
+Also note how the `NA` behavior is correct out-of-the-box with no additional adjustments.
+
+One beautiful thing about `exclude()` is that any `|` separated conditions can always be written as two sequential `exclude()` statements, meaning that you could also remove the `|` to simplify this further as:
+
+```{r}
+cars |>
+  exclude(class == "suv") |>
+  exclude(mpg < 15)
+```
+
+In general, the following form of `filter()` can always be written as a much simpler `exclude()`:
+
+```{r, eval = FALSE}
+data |> filter(!(this | that | those))
+data |> exclude(this) |> exclude(that) |> exclude(those)
```

Comment:
I appreciate how combination of `!` and `|` has a high cognitive load. But can this be written as

```r
data |>
  exclude(
    this,
    that,
    those
  )
```

This, with proper indentation, makes code much more readable, like how you would collect creation of many new variables in a single `mutate()`.